src/Deserialization.cpp

```diff
@@ -0,0 +1,1358 @@
+        auto image_param_name = deserialize_string(extern_func_argument->image_param_name());
+        Parameter image_param;
+        if (auto it = non_serialized_parameters.find(image_param_name); it != non_serialized_parameters.end()) {
+            image_param = it->second;
+        } else if (auto it = parameters_in_pipeline.find(image_param_name); it != parameters_in_pipeline.end()) {
+            image_param = it->second;
+        }
+        return ExternFuncArgument(image_param);
+    }
+}
+
+Buffer<> Deserializer::deserialize_buffer(const Serialize::Buffer *buffer) {
+    assert(buffer != nullptr);
+    if (!buffer->defined()) {
+        return Buffer<>();
+    }
+    std::string name = deserialize_string(buffer->name());
+    auto type = deserialize_type(buffer->type());
+    int32_t dimensions = buffer->dimensions();
+    std::vector<halide_dimension_t> buffer_dimensions;
+    buffer_dimensions.reserve(dimensions);
+    for (int i = 0; i < dimensions; ++i) {
+        auto dim = buffer->dims()->Get(i);
+        halide_dimension_t hl_dim;
+        hl_dim.min = dim->min();
+        hl_dim.extent = dim->extent();
+        hl_dim.stride = dim->stride();
+        buffer_dimensions.push_back(hl_dim);
+    }
+    auto fake_buffer = Buffer<>(type, nullptr, dimensions, buffer_dimensions.data(), name + "_fake");
+    auto hl_buffer = Buffer<>::make_with_shape_of(fake_buffer, nullptr, nullptr, name);
+    memcpy(hl_buffer.data(), buffer->data()->data(), buffer->data()->size());
+    return hl_buffer;
+}
+
+std::map<std::string, FunctionPtr> Deserializer::deserialize_wrapper_refs(const flatbuffers::Vector<flatbuffers::Offset<Serialize::WrapperRef>> *wrappers) {
+    assert(wrappers != nullptr);
+    std::map<std::string, FunctionPtr> result;
+    for (const auto &wrapper : *wrappers) {
+        auto name = deserialize_string(wrapper->func_name());
+        int32_t func_index = wrapper->func_index();
+        FunctionPtr func_ptr;
+        if (auto it = this->reverse_function_mappings.find(func_index); it != this->reverse_function_mappings.end() && func_index != -1) {
+            func_ptr = it->second;
+        }
+        result[name] = func_ptr;
+    }
+    return result;
+}
+
+void Deserializer::build_reverse_function_mappings(const std::vector<Function> &functions) {
+    if (!this->reverse_function_mappings.empty()) {
+        this->reverse_function_mappings.clear();
+    }
+    int cnt = 0;
+    for (const auto &f : functions) {
+        this->reverse_function_mappings[cnt++] = f.get_contents();
+    }
+}
+
+Pipeline Deserializer::deserialize(const std::string &filename) {
+    std::ifstream in(filename, std::ios::binary | std::ios::in);
+    if (!in) {
+        user_assert(false) << "failed to open file " << filename << "\n";
+        return Pipeline();
+    }
+    in.seekg(0, std::ios::end);
+    int size = in.tellg();
+    in.seekg(0, std::ios::beg);
+    std::vector<char> data(size);
+    in.read(data.data(), size);
+    in.close();
+
```

Comment:
We should probably check `in.good()` to check for errors, otherwise builds with exceptions disabled might ignore them, which would be bad